### ISSUE TITLE
Code review question for X509_NAME_print_ex method calls

### ISSUE THREAD

**Issue by IDisposable** [openssl/openssl#25209]
While reviewing the underlying code for PR https://github.com/openssl/openssl/pull/25207 I noticed a potential issue with **other** callers of that function codebase-wide.

### Background

The function `X509_NAME_print_ex(` called there either  calls down to [`do_name_ex(` in asn1/a_strex.c](https://github.com/openssl/openssl/blob/master/crypto/asn1/a_strex.c#L424) or [`X509_NAME_print(` in x509.x_name.c](https://github.com/openssl/openssl/blob/master/crypto/x509/x_name.c#L503) depending on the state of the `flags == XN_FLAG_COMPAT` compatibility mode test.

If in compatibility mode, we end up calling `X509_NAME_printf(` which returns either `1` for [success](https://github.com/openssl/openssl/blob/master/crypto/x509/x_name.c#L539) or `0` for [error](https://github.com/openssl/openssl/blob/master/crypto/x509/x_name.c#L543).

If we're not in compatibility mode we call down to [`do_name_ex(` in a_strex.c](https://github.com/openssl/openssl/blob/master/crypto/asn1/a_strex.c#L424) which will return the number of characters written. That value is initialized to `0` [here](https://github.com/openssl/openssl/blob/master/crypto/asn1/a_strex.c#L439) and then as things are emitted the `outlen` is incremented. In case of error, it will return a `-1` instead [here](https://github.com/openssl/openssl/blob/master/crypto/asn1/a_strex.c#L475) and elsewhere.

The means that the calls to `X509_NAME_print_ex(` function (and the `_fp` variant) should NOT expect any reliable values from this function other than `-1` or `0` for error, or `>=1` for any other case _(ignoring the assumed never happens path where nothing is actually emitted inside `do_name_ex(`)_ and **never** treat the return value as the number of characters emitted.

### Concern

The crux of my observation/question is that in most places _like oscp/v3_ocsp.c_ the return values is tested correctly [`<= 0`](https://github.com/openssl/openssl/blob/master/crypto/ocsp/v3_ocsp.c#L248), but in others places _like in x509/t_acert.c_ the test is only for negative values [`< 0`](https://github.com/openssl/openssl/blob/master/crypto/x509/t_acert.c#L193) _(note that the test is correct just above at line 160 [`<= 0`](https://github.com/openssl/openssl/blob/master/crypto/x509/t_acert.c#L160))_

Oddly, there are some places (like [x509/t_rec.c](https://github.com/openssl/openssl/blob/master/crypto/x509/t_req.c#L74-L76) or [x509/t_x509.c](https://github.com/openssl/openssl/blob/master/crypto/x509/t_x509.c#L137-L139) ) where extra work was done to check the output driven by the flag via code similar to:
```c
int printok = 0;

if (nmflags == XN_FLAG_COMPAT)
    printok = 1;
...
if (X509_NAME_print_ex(bp, X509_get_issuer_name(x), nmindent, nmflags) < printok)
    goto err;
```

There are also a few instances of "translation" of the return valid, presumably to clean up the oddities from an underlying layer like in [x509/x_name.c](https://github.com/openssl/openssl/blob/master/crypto/x509/x_name.c#L293-L296) where we find this:
```c
if (X509_NAME_print_ex(out, (const X509_NAME *)*pval, indent, pctx->nm_flags) <= 0)
    return 0;
return 2;
```

### Analysis

To be clear, this is a fringe case, and it's likely only applicable to error states where everything should be falling down due to errors, but the reality is that **at minimum** the code in [x509/t_acert.c](https://github.com/openssl/openssl/blob/master/crypto/x509/t_acert.c#L193) is incorrect, it should be `<= 0`

I also wonder if a slight cleanup pass to make sure that callers that test the return value are ALWAYS using/assuming `<=0` or `> 0` (and ensuring we're never expecting the value to indicate characters written). This what is *documented* to be the valid return values int the [man3 X509_NAME_print_ex.pod file](https://github.com/openssl/openssl/blob/master/doc/man3/X509_NAME_print_ex.pod?plain=1#L115-L117).

Cleaning up those uses would touch t_req.c, t_x509.c, and x_name.c.

### Not in scope

There is a lot of overlap with what is in the cloudflare-quiche submodule boringssl reference in parallel code. I am NOT suggesting we modify that stuff here, but possible to address there.

**Comment by Sashan** [comment]
I would say pull requests are always welcomed. And perhaps a better mean to discuss the matter.

The function `X509_NAME_print_ex()` is indeed unfortunate. To interpret its return value and decide
if it is error or not we need to check `nmflags` too. If `nmflags` is set to `XN_FLAG_COMPAT` (a zero),
then return value is 1, success or 0, failure.

in other cases the error is indicated by -1 (<0),  all other values indicate the number of printed
characters. seeing 0 just means there was nothing to print which is not a failure in general.

I'm going to convert this issue to discussion.
